### PR TITLE
(PUP-10375) Latest package version for pip provider fixed

### PR DESCRIPTION
**Pull request by luchihoratiu**
Sorting and comparison mechanism fixed when deciding the latest version available for a python package. Mechanism will fall through old implementation when unexpected version patterns appear as input.

### CODE REVIEW

File: lib/puppet/provider/package/pip.rb [hunk 1]
```diff
@@ -118,6 +118,15 @@ def query
     return nil
   end
 
+  def compare_pip_versions(x, y)
+    begin
+      Puppet::Util::Package::Version::Pip.compare(x, y)
+    rescue PIP_VERSION::ValidationFailure => ex
+      Puppet.debug("#{ex.message} Falling through default comparrison mechanism.")
```
Comment (ciprianbadescu) [inline lib/puppet/provider/package/pip.rb:125]:
```suggestion
      Puppet.debug("#{ex.message} Falling through default comparison mechanism.")
```

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

File: lib/puppet/util/package/version/pip.rb [hunk 2]
```diff
@@ -44,7 +44,7 @@ def self.parse(version)
 
     def self.compare(version_a, version_b)
       version_a = parse(version_a) unless version_a.is_a?(self)
-      version_b = parse(version_b) unless version_a.is_a?(self)
+      version_b = parse(version_b) unless version_b.is_a?(self)
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb:47]: Can you add a unit test for this?

File: lib/puppet/util/package/version/pip.rb [hunk 3]
```diff
@@ -150,14 +150,16 @@ def compare(this, other)
...
         raise Puppet::Error, 'Cannot compare #{this} (#{this.class}) with #{other} (Array). Only ±Float::INFINITY accepted.' unless this.abs == Float::INFINITY
-        return this <=> other.first.to_f
-      else
-        return this <=> other
+        comparison = this <=> other.first.to_f
+        return comparison unless comparison == 0
+        return -1
       end
-      0
+      this <=> other
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb:162]: are this changes covered by the new versions added on pip_spec.rb?

File: lib/puppet/provider/package/pip.rb [hunk 4]
```diff
@@ -124,35 +124,19 @@ def latest
...
-          end
-        end
-        return nil
-      end
+  def self.compare_pip_versions(x, y)
+    begin
+      Puppet::Util::Package::Version::Pip.compare(x, y)
+    rescue PIP_VERSION::ValidationFailure => ex
+      Puppet.debug("Cannot compare #{x} and #{y}. #{ex.message} Falling through default comparison mechanism.")
+      Puppet::Util::Package.versioncmp(x, y)
```
Comment (gimmyxd) [inline lib/puppet/provider/package/pip.rb:139]: Is there any test covering fallback to `Puppet::Util::Package.versioncmp`?

File: lib/puppet/util/package/version/pip.rb (hunk 3 above)
Comment (gimmyxd) [inline lib/puppet/util/package/version/pip.rb:162]: are this changes covered by the new versions added on pip_spec.rb?

File: lib/puppet/util/package/version/pip.rb (hunk 2 above)
Comment (luchihoratiu) [reply lib/puppet/util/package/version/pip.rb:47]: Added.

File: lib/puppet/provider/package/pip.rb (hunk 4 above)
Comment (luchihoratiu) [reply lib/puppet/provider/package/pip.rb:139]: Added one.

File: lib/puppet/util/package/version/pip.rb (hunk 3 above)
Comment (luchihoratiu) [reply lib/puppet/util/package/version/pip.rb:162]: The new versions added in the `versions` array of `pip_spec.rb` uncovered the need for the changes done in this commit. Those and the other versions should cover the comparison functionality.

**Comment by luchihoratiu** [comment]
jenkins please test this on centos6-64a,centos7-64a,centos8-64a

**Pull request merged**